main.c

```diff
@@ -54,87 +124,244 @@ static struct bzimage_params initialize_from_fw_cfg()
+            "Memory map conflict. ELF segment is in unusable memory (address:  0x%x, size: %d).\n",
+            elf_region.addr, elf_region.size);
+
+        if (params.cmdline_addr != 0) {
+            const struct memory_region cmdline_region = {
+                .addr = params.cmdline_addr, .size = strlen((const char *)params.cmdline_addr)};
+            die_on(memory_regions_overlap(elf_region, cmdline_region),
+                   "Memory map conflict. ELF segment overwrites command line.\n");
+        }
+
+        // Copy contents of the region from the ELF file.
+        memcpy((void *)elf_region.addr, current_elf_segment_in_memory,
+               current_program_header->filesize);
+
+        // Fill remainder of the region with zeroes (used for BSS segment).
+        memset((void *)(elf_region.addr + current_program_header->filesize), 0,
+               current_program_header->memsize - current_program_header->filesize);
+
+        current_program_header++;
+    }
+
+    // Jump to the ELF's entry point.
+    //
+    // Coreboot has already prepared the machine state specified in
+    // https://www.gnu.org/software/grub/manual/multiboot/multiboot.html#Machine-state
+    asm volatile("jmp *%[entry_ptr];" ::[entry_ptr] "r"(elf->entry),
+                 // Pass address of Multiboot information structure in EBX.
+                 "b"(mbinfo),
+                 // Report being a multiboot v1 loader in EAX.
+                 "a"(MULTIBOOT_BOOTLOADER_MAGIC));
+
+    __builtin_unreachable();
+}
+
+// Returns true if two memory regions overlap.
+bool memory_regions_overlap(const struct memory_region first, const struct memory_region second)
+{
+    const struct memory_region lower = first.addr <= second.addr ? first : second;
+    const struct memory_region higher = first.addr <= second.addr ? second : first;
+
+    // Completely disjunct regions?
+    if (higher.addr > lower.addr + lower.size) {
+        return false;
+    }
+
+    return true;
+}
+
+bool contains(const struct memory_region container, const struct memory_region containee)
+{
+    return container.addr <= containee.addr
+           && ((container.addr + container.size) >= (containee.addr + containee.size));
+}
+
+// Check whether the given memory region is within a single usable coreboot memory region.
+bool is_in_usable_coreboot_memory_region(const struct memory_region region)
+{
+    bool result = false;
+
+    for (int i = 0; i < lib_sysinfo.n_memranges; i++) {
+        struct memrange *memrange = &lib_sysinfo.memrange[i];
+        const struct memory_region coreboot_region = {.addr = memrange->base,
+                                                      .size = memrange->size};
+
+        switch (memrange->type) {
+        case CB_MEM_RAM:
+            printf("Checking RAM region 0x%lx - -0x%lx\n", coreboot_region.addr,
+                   coreboot_region.addr + coreboot_region.size);
+            if (contains(coreboot_region, region)) {
+                result = true;
```

Comment:
Can't we simply `return true;` here, and save us the result variable?

Reply:
Yes, that was a left over. It's gone.